server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -319,18 +320,26 @@ private EsExecutors() {}
         public boolean offer(E e) {
             // first try to transfer to a waiting worker thread
             if (tryTransfer(e) == false) {
-                // check if there might be spare capacity in the thread
-                // pool executor
-                int left = executor.getMaximumPoolSize() - executor.getCorePoolSize();
-                if (left > 0) {
+                // optimistically check before queueing whether there is room for more.
+                // have to use `getActiveCount` since those that are active promise to check for tasks before exiting the worker.
+                if (executor.getActiveCount() < executor.getMaximumPoolSize()) {
```

Comment:
`getActiveCount` unfortunately takes a lock and loops over all workers. We may deem this too expensive, but I'd rather have it correct unless the overhead is visible.